### ISSUE TITLE
.desktop files open as text files

### ISSUE THREAD

**Issue by RicardoEPRodrigues** [iabem97/komorebi#40]
Hello, thank you for this amazing app.

I love the added desktop icons support, yet I find it annoying when I try to open `.desktop` files (aka shortcuts) and it opens them on a text file, instead of opening the app itself.

**Comment by mjeltsch** [comment]
Yes. That sucks. Especially since I have quite a few application (some of the commercial - like PDF Studio), who only place a .desktop file to ~/Desktop to start the program. Now I have to open the command line in order to be able to start them...

**Comment by reaysawa** [comment]
Seems like the culprit line is https://github.com/cheesecakeufo/komorebi/blob/994725d147674bf7fe9392ccc1460cc41df1f8fb/src/OnScreen/Icon.vala#L176.

I'm guessing that, as it opens using the `@"file://` suffix, since .desktop files are text files anyways, it defaults to the text editor.

Some different launch strategy could be done based on the file's extension (i.e. if the file ends with **.desktop** it's a launcher, then the protocol URI prefix needs to be different).

```
 $ xdg-mime query filetype ~/Desktop/MyLauncher.desktop
application/x-desktop
```

Sadly I haven't found a suitable URI for launching applications in the URI schemes list:

https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

---

Perhaps the code is reaching for xdg-open or something similar underneath to open the files. Opening the files through xdg-open has the same effect, according to this ancient bug.

https://bugs.launchpad.net/ubuntu/+source/gvfs/+bug/378783

**Comment by reaysawa** [comment]
### I was able to fix the issue! It seems like indeed Komorebi is calling to xdg-open underneath.

---

**STEP 1.** Verify that your xdg-mime is correctly detecting the desktop launcher file as an application.

```
$ xdg-mime query filetype ~/Desktop/Launcher.desktop
application/x-desktop
```

**STEP 2.** Install dex, as suggested by Stu-axon [here](https://bugs.launchpad.net/ubuntu/+source/gvfs/+bug/378783/comments/18).

**STEP 3.** Register dex as the default application for the `application/x-desktop` mimetype following this short guide: https://superuser.com/a/309343.

For step 3 I created a file on `~/.local/share/applications/dex.desktop` with the following contents:

```ini
[Desktop Entry]
Name=Dex
Exec=dex %u
Type=Application
Terminal=false
MimeType=application/x-desktop
``` 

**STEP 4.** Final step, verify that your application/x-desktop is correctly bound to the file you created on STEP 3.

```
$ xdg-mime query default application/x-desktop
dex.desktop
```

Now clicking the desktop files on Komorebi correctly open them through Dex, which launches the application.

@cheesecakeufo Do you want me to include this solution in the repo?

**Comment by TheSandBoxMKG** [comment]
I have tried the dex workaround with no luck, the desktops files still open with Mousepad and I have followed the instructions correctly.

Debian 9 Strecth with Xfce 4.12

**Comment by chupilococo** [comment]
same issue, dex workarround dont work form me

**Comment by yanjingtu** [comment]
> ### I was able to fix the issue! It seems like indeed Komorebi is calling to xdg-open underneath.
> **STEP 1.** Verify that your xdg-mime is correctly detecting the desktop launcher file as an application.
> 
> ```
> $ xdg-mime query filetype ~/Desktop/Launcher.desktop
> application/x-desktop
> ```
> 
> **STEP 2.** Install dex, as suggested by Stu-axon [here](https://bugs.launchpad.net/ubuntu/+source/gvfs/+bug/378783/comments/18).
> 
> **STEP 3.** Register dex as the default application for the `application/x-desktop` mimetype following this short guide: https://superuser.com/a/309343.
> 
> For step 3 I created a file on `~/.local/share/applications/dex.desktop` with the following contents:
> 
> ```ini
> [Desktop Entry]
> Name=Dex
> Exec=dex %u
> Type=Application
> Terminal=false
> MimeType=application/x-desktop
> ```
> 
> **STEP 4.** Final step, verify that your application/x-desktop is correctly bound to the file you created on STEP 3.
> 
> ```
> $ xdg-mime query default application/x-desktop
> dex.desktop
> ```
> 
> Now clicking the desktop files on Komorebi correctly open them through Dex, which launches the application.
> 
> @cheesecakeufo Do you want me to include this solution in the repo's information?

========
It works. Deepin-Linux-15.11(based on Ubuntu). Steps:
```
$ sudo apt install dex
# create a file(Do the same as you do): ~/.local/share/applications/dex.desktop
$ xdg-mime query default application/x-desktop
deepin-editor.desktop
$ xdg-mime default dex.desktop application/x-desktop
$ xdg-mime query default application/x-desktop
dex.desktop
```

**Comment by joker-119** [comment]
Sorry to necro this, but the same exact issue is happening with .AppImage files on the desktop.
